magma/primitives/memory.py

```diff
@@ -0,0 +1,105 @@
+from hwtypes import BitVector
+
+from magma.bit import Bit
+from magma.bits import Bits
+from magma.bitutils import clog2
+from magma.clock import Clock, Enable
+from magma.clock_io import ClockIO
+from magma.conversions import from_bits, as_bits
+from magma.generator import Generator2
+from magma.interface import IO
+from magma.primitives.register import Register
+from magma.t import In, Out, Kind
+
+
+class CoreIRMemory(Generator2):
+    def __init__(self, depth, width, init=None):
+        self.name = f"coreir_mem{depth}x{width}"
+        addr_width = clog2(depth)
+        self.io = IO(
+            raddr=In(Bits[addr_width]),
+            rdata=Out(Bits[width]),
+            waddr=In(Bits[addr_width]),
+            wdata=In(Bits[width]),
+            clk=In(Clock),
+            wen=In(Enable)
+        )
+        self.verilog_name = "coreir_mem"
+        self.coreir_name = "mem"
+        self.coreir_lib = "coreir"
+        self.coreir_genargs = {"width": width, "depth": depth,
+                               "has_init": init is not None}
+        self.coreir_configargs = {}
+        if init is not None:
+            raise NotImplementedError("CoreIR's memory primitive does not "
+                                      "compile init arg to verilog yet")
```

Comment:
```suggestion
            raise NotImplementedError("CoreIRMemory generator does not yet "
                                      "support 'init' argument")
```

Reply:
Hmm, not quite though, it supports the argument, it just doesn't pass it through to verilog (so you can instantiate the CoreIR representation with the argument and it should compile to json file)

Reply:
Ah I see, then I would revise to "Verilog code generation for CoreIRMemory does not yet support 'init'" (not a big deal either way though, I misinterpreted the point).

Also, consider making a warning rather than exception, since the "error" is quite downstream of this entry-point. I.e. you could generate json that gets used by mapper, and the issue of verilog code gen never materializes.